### PR TITLE
Api connection

### PR DESCRIPTION
**Pull request by MeiCarmen**
Part 2 - Display Live Train Times in App

### CODE REVIEW

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 1]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+    private val stationSubmitButtonText = "View live departures"
+    private val stationSubmitButtonLoadingText = "Searching"
+
+    private val noChangesDefault = "false"
+    private val numberOfAdultsDefault = "1"
+    private val numberOfChildrenDefault = "0"
+    private val journeyTypeDefault = "single"
+    private val outboundIsArriveByDefault = "false"
+
+    private val queryOffsetInSeconds = 6*60
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: It looks like most of these are only used in one place, so can we just declare them inline there rather than set global variables. I'll add a comment to queryApiForJourneys about a nice way we could possibly handle this

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 2]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
         view.populateOriginAndDestinationSpinners(stations)
         view.setStationSubmitButtonText(stationSubmitButtonText)
     }
 
     override fun onStationSubmitButtonPressed(originStation: String, destinationStation: String) {
-        view?.openUrl("https://www.lner.co.uk/travel-information/travelling-now/live-train-times/depart/${originStation}/${destinationStation}/#LiveDepResults")
+        view?.setStationSubmitButtonText(stationSubmitButtonLoadingText)
+        launch {
+            val departureDetails = queryApiForJourneys(originStation, destinationStation)
+            view?.setDepartureTable(extractDepartureInfo(departureDetails))
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:53]: Feels weird to potentially call setDepartureTable if departure details is null. Might me nice to have code like:
```
departureDetails?.let {
    view?.setDepartureTable(extractDepartureInfo(it))
} ?: view?.handleNoDepartureInfo()
```
Where view.handleNoDepartureInfo() perhaps hides the departure table and shows the user an error

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 3]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+            view?.setStationSubmitButtonText(stationSubmitButtonText)
+        }
+    }
+
+    fun extractDepartureInfo(departureDetails: DepartureDetails?): List<DepartureInformation> {
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: If you applied my above comment you could remove the null functionality from this function, after all we probably only want to try and extract departure info from a not null object

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 4]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+
+        return departureDetails.outboundJourneys.map {
+            DepartureInformation(
+                extractSimpleTime(it.departureTime),
+                extractSimpleTime(it.arrivalTime),
+                convertToHoursAndMinutes(it.journeyDurationInMinutes),
+                it.primaryTrainOperator.name,
+                convertToPriceString(it.tickets.map { ticket -> ticket.priceInPennies }.min())
+            )
+        }
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:69]: If, as above, you removed the null handling, you could change this function into the:
`fun extractDepartureInfo(departureDetails: DepartureDetails?) = ...` format

Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:69]: You might even want to make your life simpler by passing departureDetails.outboundJourneys into the function rather than departureDetails as that's the only part you use

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 5]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+    }
+
+    fun extractSimpleTime(time: String): String {
+        val localTimeFormat = DateFormat(apiDateTimeFormat)
+        val timeWithoutTimeZone = time.split("+").first()
+        val localTime = localTimeFormat.parse(timeWithoutTimeZone)
+        return localTime.toString("HH:mm")
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:76]: I don't want to make gradle angry again but you might be able to do something like:
```
val dateTime = apiDateTimeFormat.parse(time).local
val simpleTimeFormat = DateFormat("HH:mm")
return dateTime.format(simpleTimeFormat)
```

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 6]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+                    parameters["outboundIsArriveBy"] = outboundIsArriveByDefault
+                }
+                .build()
+            return client.get<DepartureDetails> { url(url) }
+        } catch (e: Exception) {
+            println(e.toString())
+            return null
+        }
+    }
 }
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:122]: I think a lot of functions in this file can be made private

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 7]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+    fun extractSimpleTime(time: String): String {
+        val localTimeFormat = DateFormat(apiDateTimeFormat)
+        val timeWithoutTimeZone = time.split("+").first()
+        val localTime = localTimeFormat.parse(timeWithoutTimeZone)
+        return localTime.toString("HH:mm")
     }
 
+    fun convertToHoursAndMinutes(journeyDurationInMinutes: Int): String {
+        return "${journeyDurationInMinutes / 60}h ${journeyDurationInMinutes % 60}min"
+    }
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:81]: This is another function you could change into the one line format, but I'm not too fussed either way there's definitely a merit to having it like this

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 8]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+
+    fun convertToPriceString(priceInPennies: Int?): String {
+        if (priceInPennies == null) return "sold out"
+        return "£${priceInPennies / 100}.${priceInPennies % 100}"
+    }
 
+    fun getImminentDateTime(): String {
+        return (DateTime.now()
+            .add(0, queryOffsetInSeconds*1000.0))
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: Nit: Can we have spaces around the `*` please

Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: I also think this function might read clearer as:
```
fun getEarliestSearchableTime(): String {
    val dateTimeInFiveMinutes = DateTime.now().add(0, 300000)
    val apiDateTimeFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX"
    return dateTimeInFiveMinutes.format(apiDateTimeFormat)
}
```

Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: Note I believe XXX will add on the timezone

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt [hunk 9]
```diff
@@ -16,20 +19,105 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
...
+            .toString(apiDateTimeFormat) + "+00:00"
+    }
+
+    suspend fun queryApiForJourneys(
+        originStation: String,
+        destinationStation: String
+    ): DepartureDetails? {
```
Comment (JasBuc) [inline SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:102]: If you wanted to make it clear we are using defaults for some values, instead of having global variables used at the top (where they have a much wider range than needed) you could change the function call to be something like this:

```
suspend fun queryApiForJourneys(
   originStation: String,
   destinationStation: String,
   noChanges: Boolean = false,
   numberOfAdults: Int = 1
```
etc. 
This makes it clear that these are defaults we are applying when the function isn't called with a specific value. Will also mean your life is easier later when you add the user being able to alter these values.

Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:102]: If not I think it's fine to just set the default values inline, I don't think they need to be variables declared anywhere

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt [hunk 10]
```diff
@@ -33,6 +37,10 @@ class MainActivity : AppCompatActivity(), ApplicationContract.View {
         presenter.onStationSubmitButtonPressed(originStation, destinationStation)
     }
 
+    private fun setupRecyclerView() {
+        recyclerView.adapter = RecyclerAdapter()
+        recyclerView.layoutManager = LinearLayoutManager(this)
+    }
 
     override fun setStationSubmitButtonText(text: String) {
         findViewById<Button>(R.id.station_submit_button).text = text
```
Comment (JasBuc) [inline app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt]: I think you might be able to remove `findViewById<Button>(R.id.` everywhere and just call e.g. `station_submit_button.text` - lemme know if that doesn't work

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt [hunk 11]
```diff
@@ -56,4 +64,9 @@ class MainActivity : AppCompatActivity(), ApplicationContract.View {
         startActivity(intent)
     }
 
+    override fun setDepartureTable(departures: List<DepartureInformation>) {
+        (recyclerView.adapter as RecyclerAdapter).setDepartures(departures)
+        (recyclerView.adapter as RecyclerAdapter).notifyDataSetChanged()
```
Comment (JasBuc) [inline app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt]: I think we can delete this line and just call notifyDataSetChanged() at the end of the setDepartures() function

Comment (JasBuc) [reply app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt]: Also might be nice to have a global `private lateinit var recyclerAdapter: RecyclerAdapter` which we then set above, means we can just call recyclerAdapter.setDepartures..

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+
+class RecyclerAdapter : RecyclerView.Adapter<RecyclerAdapter.DepartureCellHolder>() {
+    private var departures = emptyList<DepartureInformation>()
+
+    fun setDepartures(departures: List<DepartureInformation>) {
+        this.departures = departures
+        println(this.departures)
```
Comment (JasBuc) [inline app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt]: What's the purpose of this line?

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        return DepartureCellHolder(cell)
+    }
+
+    override fun onBindViewHolder(holder: DepartureCellHolder, position: Int) {
+        holder.update(departures[position])
+    }
+
+    override fun getItemCount() = departures.size
+
+    class DepartureCellHolder(private val v: View) : RecyclerView.ViewHolder(v) {
```
Comment (JasBuc) [inline app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt]: Can we call this view rather than v please

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        fun update(departure: DepartureInformation) {
+            v.findViewById<TextView>(R.id.departure_time).text = departure.departureTime
+            v.findViewById<TextView>(R.id.arrival_time).text = departure.arrivalTime
+            v.findViewById<TextView>(R.id.price).text = departure.price
+            v.findViewById<TextView>(R.id.travel_time).text = departure.journeyTime
+            v.findViewById<TextView>(R.id.operator).text = departure.trainOperator
+            v.findViewById<TextView>(R.id.purchase).text = "Buy"
```
Comment (JasBuc) [inline app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt]: I think these can use be view.departure_time etc., might just need to say yes to some importing

File: app/src/main/res/layout/activity_main.xml [hunk 15]
```diff
@@ -1,27 +1,48 @@
...
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
-        android:layout_marginTop="32dp"
-        app:layout_constraintTop_toBottomOf="@+id/origin_station_spinner" />
+        android:orientation="vertical">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="horizontal">
```
Comment (JasBuc) [inline app/src/main/res/layout/activity_main.xml:19]: Nice use of linear layouts here 🙂

File: app/src/main/res/layout/departure_cell.xml [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+            android:layout_weight="1" />
+
+        <TextView
+            android:id="@+id/price"
+            android:layout_width="0dp"
+            android:layout_height="match_parent"
+            android:layout_marginTop="8dp"
+            android:layout_weight="1" />
+
+
```
Comment (JasBuc) [inline app/src/main/res/layout/departure_cell.xml]: Nit: Can we make this just one new line rather than two please

File: native/KotlinIOS/KotlinIOS/ViewController.swift [hunk 17]
```diff
@@ -5,18 +5,24 @@ class ViewController: UIViewController {
...
     private let presenter: ApplicationContractPresenter = ApplicationPresenter()
     
     override func viewDidLoad() {
         super.viewDidLoad()
         presenter.onViewTaken(view: self)
-        
+
+        departureTable.delegate = self
+        departureTable.dataSource = self
+        self.registerTableViewCells()
```
Comment (JasBuc) [inline native/KotlinIOS/KotlinIOS/ViewController.swift]: Could we pull these three lines out into a setUpTable function we call here and declare with the UITableViewDelegate extension section

**Review by JasBuc — COMMENTED** [review]
Looking nice! Let me know if you have any questions ☺️

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt (hunk 2 above)
Comment (MeiCarmen) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:53]: We are just about to start handling cases where we get no data or an exception and displaying it to the user. so this will be fixed there in a hopefully better manner

File: app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt (hunk 12 above)
Comment (harry48225) [reply app/src/main/java/com/jetbrains/handson/mpp/mobile/RecyclerAdapter.kt]: the println was just some debugging that we will remove

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt (hunk 8 above)
Comment (harry48225) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: I think that XXX will just add the time zone as eg "UTC" whereas we want the timezone offset "+00:00"

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt (hunk 7 above)
Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt:81]: You could also probably shorten this to just call it minutes, the function doesn't care what the minutes are referring to (and makes it easier for the function to then be reused)

File: SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt (hunk 8 above)
Comment (JasBuc) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: Hm I would've thought that to get UTC you would put "yyyy-MM-dd'T'HH:mm:ss.SSSz". It's possible the XXX gives 'Z' which I believe is equivalent to '+00:00', and I think the API won't mind, let me know if you have any issues.

Comment (harry48225) [reply SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt]: As far as we can tell XXX gives the timezone offset as "+hh:mm" in all cases when the offset is nonzero. And if the offset is 0 it gives "Z". But we've still been able to make the changes you've suggested. It wasn't working before when we were trying to convert our time first into the equivalent UTC time and then sending that to the api. Once we just sent the local time with the correct +01:00 by using the 'XXX' it worked.
